homeassistant/components/media_player/russound_rio.py

```diff
@@ -0,0 +1,267 @@
+"""
+Support for interfacing with Russound via RIO Protocol.
+
+The RIO protocol is supported by newer Russound devices such as the MCA-88.
+
+The platform will discover all enabled zones and sources and only needs to be
+provided with a name and host. It can optionally be provided with a port,
+though this should always be 9621 on any standard Russound configuration.
+
+An example configuration.yaml entry looks like this:
+
+media_player:
+  platform: russound_rio
+  host: 192.168.1.100
+  port: 9621
+  name: Russound Controller
+
+Each zone is added as a separate media player device. If the source that the
+zone is assigned to is capable of sending metadata such as artist/album name
+then this data will be reported back via the media_* properties.
+"""
+
+import asyncio
+import logging
+
+import voluptuous as vol
+
+from homeassistant.core import callback
+from homeassistant.components.media_player import (
+    SUPPORT_TURN_ON, SUPPORT_TURN_OFF, SUPPORT_VOLUME_MUTE, SUPPORT_VOLUME_SET,
+    SUPPORT_SELECT_SOURCE, MediaPlayerDevice, PLATFORM_SCHEMA,
+    MEDIA_TYPE_MUSIC)
+from homeassistant.const import (
+    CONF_HOST, CONF_PORT, STATE_OFF, STATE_ON,
+    CONF_NAME, EVENT_HOMEASSISTANT_STOP)
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['russound_rio==0.1.1']
+
+_LOGGER = logging.getLogger(__name__)
+
+SUPPORT_RUSSOUND = SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_SET | \
+                   SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_SELECT_SOURCE
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Required(CONF_NAME): cv.string,
+    vol.Optional(CONF_PORT, default=9621): cv.port,
+    })
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the Russound RIO platform."""
+    host = config.get(CONF_HOST)
+    port = config.get(CONF_PORT)
+
+    if host is None or port is None:
+        _LOGGER.error("Invalid config. Expected %s and %s",
+                      CONF_HOST, CONF_PORT)
+        return False
+
+    from russound_rio import Russound, CommandException, ZoneID
+
+    russ = Russound(hass.loop, host, port)
+
+    yield from russ.connect()
+
+    # Discover sources
+    sources = []
+    for source_id in range(1, 17):
+        try:
+            name = yield from russ.get_source_variable(source_id, 'name')
+        except CommandException:
+            break
+        if name != '':
+            sources.append((source_id, name))
```

Comment:
Move that into try block. Better would be if you API can get a list of exists device instead to trow 16 exceptions for that.